### PR TITLE
feat: added Skyra.Database scaffold

### LINKED ISSUE #1566: tracking: Skyra.Database service
The database module is the most critical component in Skyra, and we want to move it to .NET and EFCore due to the high amount of issues that arise from many different npm packages.

### Progress

- [x] Member
  - ###### Points
  - [x] `GetPoints({ UserID, GuildID }): { Experience }`
  - [x] `AddPoints({ UserID, GuildID, Points }): { Experience }`
  - [x] `RemovePoints({ UserID, GuildID, Points }): { Experience }`
  - [x] `SetPoints({ UserID, GuildID, Points }): {}`
  - [x] `ResetPoints({ UserID, GuildID }): {}`

---
- [x] User
  - ###### Points
  - [x] `GetPoints({ UserID }): { Experience }`
  - [x] `AddPoints({ UserID, Points }): { Experience }`
  - [x] `ResetPoints({ UserID }): {}`
  - ###### Money
  - [x] `AddMoney({ UserID, Money }): { Money }`
  - [x] `RemoveMoney({ UserID, Money }): { Money }`
  - [x] `SetMoney({ UserID, Money }): {}`
  - [x] `GiveMoney({ UserID, TargetID, Money }): { AuthorMoney, TargetMoney }`
  - ###### Claim
  - [x] `ClaimDaily({ UserID, Force }): { Remaining }`
    - **Note**: It updates `UserCooldown.Daily`.
    - **Note**: `Force` runs the command if Remaining is less than 1 hour. Otherwise, it should return `Success = false`.
  - ###### Reputation
  - [x] `GiveReputation({ UserID, TargetID }): {}`
    - **Note**: It updates `UserCooldown.Reputation`.
  - ###### Banners (Profile)
  - [x] `GetLargeBanners({ UserID }): { string[] }`
  - [x] `SetLargeBanner({ UserID }): {}`
  - [x] `GetAllLargeBanners({ UserID }): { string[] }`
  - ###### Banners (Level)
  - [x] `GetSmallBanners({ UserID }): { string[] }`
  - [x] `SetSmallBanner({ UserID }): {}`
  - [x] `GetAllSmallBanners({ UserID }): { string[] }`
  - ###### Badges
  - [x] `GetBadges({ UserID }): { Badges }`
  - [x] `GetAllBadges({ UserID }): { Badges }`
  - ###### Profile
  - [x] `GetColor({ UserID }): { Color }`
  - [x] `GetProfile({ UserID }): { Experience, Color, Reputation, Money, Vault, Badges, BannerProfile }`
  - ###### Spouses
  - [x] `GetSpouses({ UserID }): UserID[]`
  - [x] `AddSpouse({ UserID, TargetID }): {}`
  - [x] `RemoveSpouse({ UserID, TargetID }): {}`
  - ###### Game Integration
  - [x] `GetGameIntegration({ UserID, Game }): { JSON? }`
  - [x] `AddGameIntegration({ UserID, Game, JSON }): {}`
  - [x] `RemoveGameIntegration({ UserID, Game }): { JSON? }`
  - ###### Account Management
  - [x] `Remove({ UserID }): {}`
    - **Note**: Deletes the `User`, `UserCooldown`, and `UserProfile` entries if existing.
    - **Note**: This will come handy for removing all data from a deleted user, in accordance to GDPR.
    - Also useful for nuking data before unit tests.

---
- [x] Social
  - ###### Banners
  - [x] `GetBanners(): Banner[]`
  - [x] `AddBanner(Banner): {}`
  - [x] `UpdateBanner(Banner): {}`
  - [x] `RemoveBanner(BannerID): {}`
  - ###### Leaderboard
  - [x] `GetLocalLeaderboard({ GuildID }): UserID[]`
  - [x] `GetGlobalLeaderboard(): UserID[]`

---
- [x] Guild
  - [x] `Get({ GuildID }): { ... }`
  - [x] `Update({ GuildID, JSON }): JSON`
    - **Note**: Takes a partial JSON object with the properties to update.
    - **Note**: Returns the patched data.
  - [x] `Remove({ GuildID }): {}`

---
- [x] Starboard
  - [x] `AddMessage(StarboardData): { StarMessageID?, Count }`
  - [x] `RemoveMessage({ MessageID, ChannelID }): { StarMessageID? }`
  - [x] `RemoveMessages({ MessageID[], ChannelID }): { StarMessageID[] }`
  - [x] `RemoveChannel({ ChannelID }): { StarMessageID[] }`

---
- [x] Schedule
  - [x] `Add(TaskData): {}`
  - [x] `GetAll(): TaskData[]`
  - [x] `Update(TaskData[]): {}`
  - [x] `Remove(TaskID[]): {}`
  - [x] `RemoveAll(): {}`

---
- [x] Giveaway
  - [x] `Add(GiveawayData): {}`
  - [x] `GetAll(): GiveawayData[]`
  - [x] `Remove({ MessageID, ChannelID }): {}`

---
- [x] Runner
  - [x] `Run({ Query }): JSON`
    - **Note**: Runs arbitrary SQL queries, and returns a dynamic JSON object.

> **Note**: This is not an exhaustive list, we will find commands we will need as we go.

### PR DESCRIPTION
**Pull request by kyranet**

### CODE REVIEW

**Review by Tylertron1998 — COMMENTED** [review]
wasn't there the idea of moving all services into a /services folder, to avoid cluttering up everything with .sln and what not?

File: .github/workflows/continuous-integration.yml [hunk 1]
```diff
@@ -87,6 +87,13 @@ jobs:
         with:
           name: coverage
           path: coverage/
+      - name: Setup dotnet
+        uses: actions/setup-dotnet@v1
+        with:
+          dotnet-version: '5.0.103'
+      - name: Run dotnet tests
+        run: |
+          dotnet test services/Skyra.Tests
```
Comment (favna) [inline .github/workflows/continuous-integration.yml]:
```suggestion
        run: dotnet test services/Skyra.Tests
```

Please don't do this. You only need a pipe for multi-line scripts.

File: .github/workflows/continuous-integration.yml [hunk 2]
```diff
@@ -114,6 +121,13 @@ jobs:
           popd
       - name: Build Code
         run: yarn build
+      - name: Setup dotnet
+        uses: actions/setup-dotnet@v1
+        with:
+          dotnet-version: '5.0.103'
+      - name: Build Services
+        run: |
+          dotnet build services
```
Comment (favna) [inline .github/workflows/continuous-integration.yml]:
```suggestion
        run: dotnet build services
```

**Review by favna — CHANGES_REQUESTED** [review]
Also will we be moving the bot stuff to its own folder in this PR? If yes please do it at the very very end and once the PR has been approved by reviewers. Otherwise, a PR right after this one.

**Comment by Tylertron1998** [comment]
> Also will we be moving the bot stuff to its own folder in this PR? If yes please do it at the very very end and once the PR has been approved by reviewers. Otherwise, a PR right after this one.

I'm happy for moving the bot over into it's own repo - would make everything a *lot* cleaner imo. But might be a little more suitable in it's own PR? As I imagine a lot of pathing in commands and stuff (image commands for example) will need to be changed.

**Comment by favna** [comment]
@Tylertron1998 Not its own repo. Heck if anything goes in another repo it would be this, not the bot. This is and will always be the bot's repo. No instead it would go into it's own folder.

I meant something like this:

```
\skyra
├── crowdin.yml
├── docker
|  ├── control.ps1
|  ├── docker-compose.yml
|  └── README.md
├── LICENSE
├── README.md
├── services
|  ├── Directory.Build.props
|  ├── Skyra.Database
|  ├── Skyra.sln
|  ├── Skyra.sln.DotSettings
|  └── Skyra.Tests
└── skyra
   ├── assets
   ├── jest.config.ts
   ├── ormconfig.js
   ├── package.json
   ├── pm2.ecosystem.yml
   ├── scripts
   ├── src
   ├── tests
   ├── tsconfig.base.json
   ├── tsconfig.eslint.json
   └── yarn.lock

directory: 9 file: 16
```

image

**Comment by kyranet** [comment]
Yes, `skyra/skyra`.

**Comment by favna** [comment]
@kyranet `skyra/bot` seems derp as well lol... I have no idea what else we can do.

**Comment by kyranet** [comment]
`skyra/main`, `skyra/worker`, `skyra/services/Skyra.Main`, pick your weapon.

**Comment by Tylertron1998** [comment]
whoops, I meant in it's own folder, lol

**Comment by Tylertron1998** [comment]
`skyra/bot`? because.. you know.. it's where the 'bot' part of skyra is? :P

**Comment by Stitch07** [comment]
+1 for skyra/bot

**Comment by Tylertron1998** [comment]
> @Tylertron1998 Not its own repo. Heck if anything goes in another repo it would be this, not the bot. This is and will always be the bot's repo. No instead it would go into it's own folder.
> 
> I meant something like this:
> 
> ```
> \skyra
> ├── crowdin.yml
> ├── docker
> |  ├── control.ps1
> |  ├── docker-compose.yml
> |  └── README.md
> ├── LICENSE
> ├── README.md
> ├── services
> |  ├── Directory.Build.props
> |  ├── Skyra.Database
> |  ├── Skyra.sln
> |  ├── Skyra.sln.DotSettings
> |  └── Skyra.Tests
> └── skyra
>    ├── assets
>    ├── jest.config.ts
>    ├── ormconfig.js
>    ├── package.json
>    ├── pm2.ecosystem.yml
>    ├── scripts
>    ├── src
>    ├── tests
>    ├── tsconfig.base.json
>    ├── tsconfig.eslint.json
>    └── yarn.lock
> 
> directory: 9 file: 16
> ```
> 
> image

scratch that, this looks perfect, favna 4 the win

**Pull request closed without merging**